test/integration/user-management-api/application_plan_features_test.rb

```diff
@@ -105,10 +105,11 @@ def teardown
 
     test 'associate existing feature' do
       feat = FactoryBot.create(:feature, featurable: @provider.default_service)
+      feat2 = FactoryBot.create(:feature, featurable: @provider.default_service)
 
       assert_difference @app_plan.features.method(:count), 2 do
         post admin_api_application_plan_features_path(@app_plan, format: :xml), params: params.merge({ feature_id: feat.id })
-        post admin_api_application_plan_features_path(@app_plan, format: :xml), params: params.merge({ feature_id: feat.id })
+        post admin_api_application_plan_features_path(@app_plan, format: :xml), params: params.merge({ feature_id: feat2.id })
```

Comment:
Looking at the original test - does it mean that you could "attach" the same feature to the plan twice, and it would look like two features?... :thinking: 

So, this would mean 2 entries in the table with the exact same columns?

Reply:
I don't know how would it look like from UI, but yes, two repeated rows in the table. That's why the test failed in this branch, because of the PK.

Reply:
I checked this, and the UI is immune to this duplication, but API does return repeated entries. This certainly looks like a bug to me... 

I think we probably need to include this change in the release notes for on-premises, and also make support aware of this change, so they can assist customers whose migrations might fail because of this (I hope there will be no issues, but who knows... )

Reply:
Example error on migration:
```
$ SAFETY_ASSURED=1 bundle exec rails db:migrate:up VERSION=20230719112703
Removing :environment prerequisite from db:create
== 20230719112703 AddPrimaryKeyToFeaturesPlans: migrating =====================
-- remove_index(:features_plans, [:plan_id, :feature_id])
   -> 0.0163s
-- execute("ALTER TABLE features_plans ADD PRIMARY KEY (plan_id, feature_id)")
rails aborted!
StandardError: An error has occurred, all later migrations canceled:

Mysql2::Error: Duplicate entry '8-23' for key 'PRIMARY': ALTER TABLE features_plans ADD PRIMARY KEY (plan_id, feature_id)
/home/dmayorov/Projects/3scale/porta/db/migrate/20230719112703_add_primary_key_to_features_plans.rb:6:in `up'
bin/rails:4:in `require'
bin/rails:4:in `<main>'

Caused by:
ActiveRecord::RecordNotUnique: Mysql2::Error: Duplicate entry '8-23' for key 'PRIMARY': ALTER TABLE features_plans ADD PRIMARY KEY (plan_id, feature_id)
/home/dmayorov/Projects/3scale/porta/db/migrate/20230719112703_add_primary_key_to_features_plans.rb:6:in `up'
bin/rails:4:in `require'
bin/rails:4:in `<main>'

Caused by:
Mysql2::Error: Duplicate entry '8-23' for key 'PRIMARY'
/home/dmayorov/Projects/3scale/porta/db/migrate/20230719112703_add_primary_key_to_features_plans.rb:6:in `up'
bin/rails:4:in `require'
bin/rails:4:in `<main>'
Tasks: TOP => db:migrate:up
(See full trace by running task with --trace)
```

Reply:
Agreed. I'll work on a query or script to hand to clients if they run into this problem. Fortunately, there are no repeated records in SaaS, which is a good indicator the code is working properly.

Reply:
Great, thank you!

Reply:
I'm adding this to the release notes, let me know if you have some comments:

----
The migration could fail due to duplicate entries in the `features_plans` table. These are the errors raised by the DB engine in this case:

MySQL:
`Mysql2::Error: Duplicate entry '8-23' for key 'PRIMARY': ALTER TABLE features_plans ADD PRIMARY KEY (plan_id, feature_id)`

Postgres:
```
PG::UniqueViolation: ERROR:  could not create unique index "features_plans_pk"
DETAIL:  Key (plan_id, feature_id)=(8, 3) is duplicated.
: ALTER TABLE features_plans ADD CONSTRAINT features_plans_pk PRIMARY KEY (plan_id, feature_id)
```

Oracle:
`OCIError: ORA-02437: cannot validate (RAILS.SYS_C0010658) - primary key violated: ALTER TABLE FEATURES_PLANS ADD PRIMARY KEY (PLAN_ID, FEATURE_ID)`

The duplicated rows must be removed from the table before retrying to run the migration. The next query returns a list of duplicated rows (Valid for three DBs):
	
```
SELECT plan_id, feature_id, count(*)
FROM features_plans
GROUP BY plan_id, feature_id
HAVING count(*) > 1
```

When duplicated rows are found, we must keep one row for each of the results returned by the query above, and delete it's duplicates. The process to delete the duplicated rows is different depending on the DB engine and version. For instance, one can use "**ROW_NUM**" method for oracle or MySQL 8+, but only the "**Intermediate table**" method is available for MySQL 5.7. The next links summarize all possible situations:

MySQL:
https://www.mysqltutorial.org/mysql-delete-duplicate-rows/

PostgreSQL:
https://www.postgresqltutorial.com/postgresql-tutorial/how-to-delete-duplicate-rows-in-postgresql/

Oracle:
https://www.oracletutorial.com/advanced-oracle-sql/how-to-delete-duplicate-records-in-oracle/

In the worst scenario when removing duplicates isn't possible, this migration can be safely ignored unless the client wants to use "**Group Replication**" feature in MySQL.


Additionally, in Oracle the next error can happen when retrying the migration after failing because of duplicates:

`No indexes found on features_plans with the options provided.`

This can be solved by running the next SQL query and trying the migration again

`CREATE INDEX FEATURES_PLANS_PLAN_ID_IDX ON FEATURES_PLANS (PLAN_ID,FEATURE_ID);`

Reply:
done!